src/cockpit/channels/filesystem.py

```diff
@@ -268,3 +287,226 @@ def do_close(self):
         self._watch.close()
         self._watch = None
         self.close()
+
+
+class FsInfoChannel(Channel):
+    payload = 'fsinfo'
+
+    # Options (all get set in `do_open()`)
+    path: str
+    attrs: 'set[str]'
+    fnmatch: str
+    follow: bool
+    watch: bool
+
+    # State
+    current_value: JsonDict
+    effective_fnmatch: str = ''
+    fd: 'Handle | None' = None
+    pending: 'set[str] | None' = None
+    path_watch: 'PathWatch | None' = None
+    getattrs: 'Callable[[int, str], JsonDocument]'
+
+    @staticmethod
+    def make_getattrs(attrs: Sequence[str]) -> 'Callable[[int, str], JsonDocument | None]':
+        # Cached for the duration of the closure we're creating
+        @functools.lru_cache()
+        def get_user(uid: int) -> 'str | int':
+            try:
+                return pwd.getpwuid(uid).pw_name
+            except KeyError:
+                return uid
+
+        @functools.lru_cache()
+        def get_group(gid: int) -> 'str | int':
+            try:
+                return grp.getgrgid(gid).gr_name
+            except KeyError:
+                return gid
+
+        stat_types = {stat.S_IFREG: 'reg', stat.S_IFDIR: 'dir', stat.S_IFLNK: 'lnk', stat.S_IFCHR: 'chr',
+                      stat.S_IFBLK: 'blk', stat.S_IFIFO: 'fifo', stat.S_IFSOCK: 'sock'}
+        available_stat_getters = {
+            'type': lambda buf: buf and stat_types.get(stat.S_IFMT(buf.st_mode)),
+            'tag': tag_from_stat,
+            'mode': lambda buf: stat.S_IMODE(buf.st_mode),
+            'size': lambda buf: buf.st_size,
+            'uid': lambda buf: buf.st_uid,
+            'gid': lambda buf: buf.st_gid,
+            'mtime': lambda buf: buf.st_mtime,
+            'user': lambda buf: get_user(buf.st_uid),
+            'group': lambda buf: get_group(buf.st_gid),
+        }
+        stat_getters = tuple((key, available_stat_getters.get(key, lambda _: None)) for key in attrs)
+
+        def get_attrs(fd: int, name: str) -> 'JsonDict | None':
+            try:
+                buf = os.lstat(name, dir_fd=fd) if name else os.fstat(fd)
+            except FileNotFoundError:
+                return None
+            except OSError:
+                return {name: None for name, func in stat_getters}
+
+            result = {key: func(buf) for key, func in stat_getters}
+
+            if 'target' in result and stat.S_IFMT(buf.st_mode) == stat.S_IFLNK:
```

Comment:
If we don't support a particular attribute, we still report it as None (due to the way we construct the `stat_getters` — if we don't know the attribute, we return None.  So the keys of `result` are exactly the requested attributes.  Checking a dictionary instead of checking a list (which is maybe more logical) of a pure performance thing.

Note: having the None values here helps us because it means we don't need to (for example) explicitly set 'target' to None in case of the file not being a symlink or the syscall failing.  That just keeps things a bit tidier below.

Reply:
Ah, now I get it -- you need to explicitly ask for `target` in `attrs`.